algorithms/rezzuf/rezzuf_hierarflow_routines.cpp

```diff
@@ -0,0 +1,237 @@
+/*
+ * fuzzuf
+ * Copyright (C) 2022 Ricerca Security
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see http://www.gnu.org/licenses/.
+ */
+
+#include "fuzzuf/algorithms/rezzuf/rezzuf_hierarflow_routines.hpp"
+
+#include <limits>
+
+#include "fuzzuf/algorithms/aflplusplus/aflplusplus_havoc.hpp"
+#include "fuzzuf/algorithms/aflplusplus/aflplusplus_util.hpp"
+#include "fuzzuf/algorithms/rezzuf/rezzuf_state.hpp"
+#include "fuzzuf/algorithms/rezzuf/rezzuf_testcase.hpp"
+
+namespace fuzzuf::algorithm::rezzuf::routine {
+
+// FIXME: some of Rezzuf's HierarFlowRoutines such as RezzufSelectSeed and those
+// of AFL++ are almost the same. These should be unified into one template
+// struct in future refactoring. The unification requires a bit large diff due
+// to the following reasons:
+// 1. AFL++ reuses `afl::BuildAFLFuzzFlow` in building its fuzz_loop by
+// overriding some HierarFlowRoutines of AFL with explicit specialization.
+// 2. AFL++ and Rezzuf use different State structs: AFLplusplusState and
+// RezzufState. And there is no inheritance between them. To reuse AFL++'s
+// HierarFlowRoutines in Rezzuf, they should be defined as independent structs,
+// not as the explicit specialization of AFL. To this end, first AFL++ should
+// stop using `afl::BuildAFLFuzzFlow` and define its own `BuildFuzzFlow`
+// function.
+
+RezzufSelectSeed::RezzufSelectSeed(RezzufState &state) : state(state) {}
+
+utils::NullableRef<hierarflow::HierarFlowCallee<void(void)>>
+RezzufSelectSeed::operator()(void) {
+  static size_t runs_in_current_cycle = std::numeric_limits<size_t>::max();
+
+  if (state.queue_cycle == 0 ||
+      runs_in_current_cycle > state.case_queue.size()) {
+    state.queue_cycle++;
+    state.current_entry = state.seek_to;
```

Comment:
This is a mess that is historically brought from AFL/AFL++ . 
AFL selects entries in a round-robin manner (therefore `seek_to` decides the start index), while AFL++ usually does randomlly with a probability distribution. Since rezzuf does not need to support the old selection manner, we can delete these part as you said.

Reply:
Fixed in https://github.com/fuzzuf/fuzzuf/pull/92/commits/3b147b2f027923257aab1c0a40faf392d242ce5e